### PR TITLE
Added solutions to exercises

### PR DESCRIPTION
**Pull request by leandrotula**

### CODE REVIEW

File: tp1/tp1.go [hunk 1]
```diff
@@ -24,17 +27,112 @@ type Carrito struct {
 // para cada super mercado, cuanto saldria comprar esos productos ahi.
 // Retorna un slice de carritos, donde se tiene uno para cada super mercado.
 func (p Productos) CalcularPrecios(ids ...int) []Carrito {
-	return nil
+
+	all, error := LeerProductos("productos.json")
```
Comment (matipan) [inline tp1/tp1.go]: Esto no es necesario. Los archivos de test carga los productos en memoria y te dan como input lo que necesitas. Si te fijas, `CalcularPrecios` es un metodo del tipo `Productos`, y productos es un slice de slices de string. Con iterar `p` ya vas a poder obtener el input acorde. Ademas de eso, los tests tienen otros escenarios con otro tipo de data para asegurarnos que se estan considerando edge cases

File: tp1/tp1.go [hunk 2]
```diff
@@ -24,17 +27,112 @@ type Carrito struct {
...
+	var collectedResult map[string]int
+
+	if error != nil {
+		panic("Problemas durante apertura de archivo")
+	}
+
+	for _, singleProduct := range all {
+
+		for _, marketID := range ids {
+			foundID := convertStringToInt(singleProduct[1])
```
Comment (matipan) [inline tp1/tp1.go]: Los indices de cada atributo en el arreglo podrian ser constantes, incluso podriamos usar algo como `iota` para que sean autoincrementales:
```
const (
    SUPERMERCADO = iota
    ID_PRODUCTO
    PRECIO_PRODUCTO
```

File: tp1/tp1.go [hunk 3]
```diff
@@ -24,17 +27,112 @@ type Carrito struct {
...
+			if foundID == marketID {
```
Comment (matipan) [inline tp1/tp1.go]: en go se suele manejar los casos al reves, primero se maneja el "sad path" que te genera una condicion de corte o de seguir iterando. En este caso podrias tener algo asi:
```
if foundID != marketID {
    continue
}
// codigo que maneja el caso donde se encuentre
```

File: tp1/tp1.go [hunk 4]
```diff
@@ -24,17 +27,112 @@ type Carrito struct {
...
+
+	return pricesResult
 }
 
 // Promedio recibe el id de un producto y retorna el precio promedio
 // de ese producto usando los precios de todos los supermercados.
 func (p Productos) Promedio(idProducto int) float64 {
-	return 0
+
+	all, error := LeerProductos("productos.json")
```
Comment (matipan) [inline tp1/tp1.go]: Mismo caso que antes. Productos ya estan cargados en `p`

File: tp1/tp1.go [hunk 5]
```diff
@@ -24,17 +27,112 @@ type Carrito struct {
...
+	return float64(sumAllPrices) / float64(dataLeng)
 }
 
 // BuscarMasBarato recibe un id de producto a buscar y te retorna
 // el producto mas barato que haya encontrado.
 func (p Productos) BuscarMasBarato(idProducto int) (Producto, bool) {
-	return nil, false
+	all, error := LeerProductos("productos.json")
+
+	var tmpValue int = 4294967295
```
Comment (matipan) [inline tp1/tp1.go]: Para el maximo podrias usar `math.MaxInt64`

File: tp1/tp1.go [hunk 6]
```diff
@@ -24,17 +27,112 @@ type Carrito struct {
...
+			found = true
+			priceConverted := convertStringToInt(value[2])
+
+			if priceConverted < tmpValue {
+				tmpValue = priceConverted
+			}
+		}
+	}
+
+	if !found {
```
Comment (matipan) [inline tp1/tp1.go]: Es necesario este if? Si no fue encontrado found va a ser false y precio va a ser 0, no? Podria simplificar un poquito

File: tp1/tp1.go [hunk 7]
```diff
@@ -7,6 +7,9 @@ type Producto interface {
 	Precio() int
 }
 
+// Estructura que implementara ambos metodos ID y Precio
+type productImplementation []int
```
Comment (matipan) [inline tp1/tp1.go]: Aca seria preferible usar algo como una estructura que facilite un poco la implementacion, no? Podrias definir alguna estructura con los campos necesarios y que los metodos actuen de "getters", por asi decirle.

**Comment by leandrotula** [comment]
@matipan muchas. gracias por tu revision !!!! A seguir aprendiendo!
